eo-phi-normalizer/src/Language/EO/Phi/Rules/Yaml.hs

```diff
@@ -72,12 +80,26 @@ parseRuleSetFromFile = Yaml.decodeFileThrow
 convertRule :: Rule -> Common.Rule
 convertRule Rule{..} ctx obj =
   [ obj'
-  | subst <- matchObject pattern obj
+  | contextSubsts <- matchContext ctx obj context
+  , let pattern' = applySubst contextSubsts pattern
+  , let result' = applySubst contextSubsts result
+  , subst <- matchObject pattern' obj
   , all (\cond -> checkCond ctx cond subst) when
-  , obj' <- [applySubst subst result]
+  , obj' <- [applySubst subst result']
   , not (objectHasMetavars obj')
   ]
 
+matchContext :: Common.Context -> Object -> Maybe RuleContext -> [Subst]
+matchContext Common.Context{..} obj = \case
+  Nothing -> [emptySubst]
+  Just (RuleContext Nothing Nothing) -> [emptySubst]
+  Just (RuleContext (Just pattern) Nothing) -> matchObject pattern globalObject
+  Just (RuleContext Nothing (Just pattern)) -> matchObject pattern thisObject
+  Just (RuleContext (Just globalPattern) (Just thisPattern)) -> matchObject globalPattern globalObject ++ matchObject thisPattern thisObject
```

Comment:
This part can most definitely be written in a cleaner way, but I wanted to verify the logic first.